Optimise Random.Shuffle

hamarb123 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/119860

This change does the following:
- Remove the `i != j` check
- Remove unnecessary local `n` (which caused extra assembly)

Benchmarks on AMD CPU (thanks EgorBot):
<img width="573" height="519" alt="image" src="https://github.com/user-attachments/assets/604e587d-64e8-4478-b9e0-26957e8e33df" />

hamarb123 committed:
Update Random.cs

Implement optimisation from https://github.com/dotnet/runtime/issues/119860 for structs of size 16 or less.

xtqqczze commented:
> not sure if we prefer Unsafe.SizeOf or sizeof+unsafe

`sizeof` is preferred over `Unsafe.SizeOf`, see https://github.com/dotnet/runtime/pull/78741

xtqqczze commented:
I think you'll get more efficient register allocation by inlining the local `n`.

hamarb123 committed:
Implement feedback

hamarb123 committed:
Make comment spacing nicer

jkotas commented:
@EgorBo Could you please check whether https://github.com/dotnet/runtime/pull/119890#issuecomment-3313739047 got stuck?

xtqqczze commented:
> > I think you'll get more efficient register allocation by inlining the local `n`.
> 
> Not sure what you're meaning by this.

I mean replace the variable `n` with `values.Length`

hamarb123 committed:
Update Random.cs

Move value of `n` (values.Length) inline to improve register allocation.

xtqqczze commented on src/libraries/System.Private.CoreLib/src/System/Random.cs:
```diff
@@ -361,15 +361,17 @@ public void Shuffle<T>(T[] values)
+                // cost of the read and write when the write size is small.
+                // So, for all reference types and for all small structs,
+                // do the write regardless.
+                if (sizeof(T) <= 16 || j != i)
```
```suggestion
                if (sizeof(T) <= 2 * sizeof(nint) || j != i)
```

I don't know whether a threshold of 16 will cause regressions on 32-bit, so this seems sensible.

hamarb123 replied:
Ah yeah, forgot about arm32 lol 😅

xtqqczze replied:
What about using `Vector<byte>.Count`?

hamarb123 committed:
Update Random.cs

Change limit to 2*pointer rather than 16 (still 16 on 64-bit).

jkotas commented:
The core libraries are not optimized for large structs that do not follow guidelines https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct. To shuffle span of large structs in most efficient way, you would want to use an algorithm that does ~N element writes, not ~2N element writes like the current algorithm. Providing second implementation like that in the core libraries would be over-engineering inconsistent with the general approach.

There are two variables in play: cost of mis-predicted branch (this varies a lot between hardware) and cost of reading and writing an element of a specific type (this depends on more than just the size). The cost of the mis-predicted branch grows as the amount to shuffle shrinks. The branch is 50/50 coinflip in the last iteration of the loop. I think that it is hard to come up with a simple formula that approximates comparison of these two costs and holds well across broad variety of hardware. It may be best to just delete the mis-predicted branch to keep this simple.

You may want to loop at other tweaks to make this method faster. For example, call `Next(int maxValue)` instead of the two-argument Next overload.

hamarb123 commented:
> It may be best to just delete the mis-predicted branch to keep this simple.

Yeah cool, I'll just delete the check entirely then.

> You may want to look at other tweaks to make this method faster. For example, call `Next(int maxValue)` instead of the two-argument Next overload.

This didn't seem to be faster (although I didn't try inverting the loop order in its entirety). However, I tested unrolling it 2x like so:
```csharp
for (int i = 0; i < values.Length - 2; i += 2)
{
	int j1 = r.Next(i, values.Length);
	int j2 = r.Next(i + 1, values.Length);
	T temp1 = values[i];
	T temp2 = values[i + 1];
	values[i] = values[j1];
	values[i + 1] = values[j2];
	values[j1] = temp1;
	values[j2] = temp2;
}
```
and it did seem to provide a significant performance advantage on small input sizes (about 2x on the suitably small ones). However this does change behaviour for the `new Random(seed)` version - not sure how problematic this is (lmk if it's probkematic).

I might add these changes & then run an egorbot benchmark to compare all of the changes so far to the original.

Also, can you add tenet-performance to this PR? Thanks :)

hamarb123 committed:
Update Random.cs

- Unconditionally remove the i != j check
- Unroll loop by 2 to improve performance of small cases in particular

hamarb123 committed:
Update Random.cs

- Remove now-unnecessary unsafe

hamarb123 commented:
@EgorBot -amd
```csharp
using BenchmarkDotNet.Attributes;

public class Bench
{
    public static string[] s_a = GenerateStrings();

    static string[] GenerateStrings()
    {
         var a = new string[8];
         for (int i = 0; i < a.Length; i++) a[i] = i.ToString();
         return a;
    }

    public static int[] i_a = GenerateInts();

    static int[] GenerateInts()
    {
         var a = new int[8];
         for (int i = 0; i < a.Length; i++) a[i] = i;
         return a;
    }

    [Benchmark]
    public void ShuffleStrings() => Random.Shared.Shuffle(s_a);

    [Benchmark]
    public void ShuffleInts() => Random.Shared.Shuffle(i_a);
}
```

hamarb123 commented:
@EgorBot -amd
```csharp
// This one is of size 4 - last one was size 8

using BenchmarkDotNet.Attributes;

public class Bench
{
    public static string[] s_a = GenerateStrings();

    static string[] GenerateStrings()
    {
         var a = new string[4];
         for (int i = 0; i < a.Length; i++) a[i] = i.ToString();
         return a;
    }

    public static int[] i_a = GenerateInts();

    static int[] GenerateInts()
    {
         var a = new int[4];
         for (int i = 0; i < a.Length; i++) a[i] = i;
         return a;
    }

    [Benchmark]
    public void ShuffleStrings() => Random.Shared.Shuffle(s_a);

    [Benchmark]
    public void ShuffleInts() => Random.Shared.Shuffle(i_a);
}
```

hamarb123 commented:
@EgorBot -amd
```csharp
// This one is of size 2 - last one was size 4

using BenchmarkDotNet.Attributes;

public class Bench
{
    public static string[] s_a = GenerateStrings();

    static string[] GenerateStrings()
    {
         var a = new string[2];
         for (int i = 0; i < a.Length; i++) a[i] = i.ToString();
         return a;
    }

    public static int[] i_a = GenerateInts();

    static int[] GenerateInts()
    {
         var a = new int[2];
         for (int i = 0; i < a.Length; i++) a[i] = i;
         return a;
    }

    [Benchmark]
    public void ShuffleStrings() => Random.Shared.Shuffle(s_a);

    [Benchmark]
    public void ShuffleInts() => Random.Shared.Shuffle(i_a);
}
```

hamarb123 committed:
Update Random.cs

Remove (incorrectly) unrolled logic, and just do it one at a time like before.

hamarb123 commented:
@EgorBot -amd
```csharp
using BenchmarkDotNet.Attributes;

public class Bench
{
    [Params(2, 4, 8)]
    public int ArraySize { get; set; }

    public string[] s_a = null!;

    [GlobalSetup]
    public void GlobalSetup()
    {
        s_a = GenerateStrings(ArraySize);
        i_a = GenerateInts(ArraySize);
    }

    static string[] GenerateStrings(int length)
    {
         var a = new string[length];
         for (int i = 0; i < a.Length; i++) a[i] = i.ToString();
         return a;
    }

    public int[] i_a = null!;

    static int[] GenerateInts(int length)
    {
         var a = new int[length];
         for (int i = 0; i < a.Length; i++) a[i] = i;
         return a;
    }

    [Benchmark]
    public void ShuffleStrings() => Random.Shared.Shuffle(s_a);

    [Benchmark]
    public void ShuffleInts() => Random.Shared.Shuffle(i_a);
}
```

hamarb123 committed:
Update Random.cs

hamarb123 committed:
Update Random.cs

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Random.cs:
```diff
@@ -363,18 +363,16 @@ public void Shuffle<T>(T[] values)
+                int j = Next(i, values.Length);
+
+                // The i != j check is excluded intentionally.
+                // Microbenchmarks show that the extra branch costs more than the redundant read/write for small value types.
```
```suggestion
                // Microbenchmarks show that the mispredicted branches cost more than the redundant read/write for small value types.
```

hamarb123 committed:
Update Random.cs
